File: bundle/jsky.app.ot/src/main/java/jsky/app/ot/gemini/editor/targetComponent/GuidingFeedbackEditor.java
package jsky.app.ot.gemini.editor.targetComponent;

import edu.gemini.ags.api.AgsMagnitude;
import edu.gemini.shared.util.immutable.Option;
import edu.gemini.spModel.obs.context.ObsContext;
import edu.gemini.spModel.target.SPTarget;
import jsky.app.ot.OT;

import java.awt.*;

/**
 * Displays the guiding feedback associated with a target, if any.
 */
final class GuidingFeedbackEditor implements TelescopePosEditor {
    private final GuidingFeedback.Table tab = new GuidingFeedback.Table();

    public Component getComponent() { return tab.peer(); }

    @Override
    public void edit(final Option<ObsContext> ctxOpt, final SPTarget target) {
        final AgsMagnitude.MagnitudeTable mt = OT.getMagnitudeTable();
        if (ctxOpt.isEmpty()) {
            tab.clear();
        } else {
            tab.showRows(GuidingFeedback.targetAnalysis(ctxOpt.getValue(), mt, target));
        }
    }
}

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/GuidingFeedbackEditor.scala
package jsky.app.ot.gemini.editor.targetComponent

import java.awt.Component

import edu.gemini.pot.sp.ISPNode
import edu.gemini.shared.util.immutable.{Option => GOption}
import edu.gemini.spModel.obs.context.ObsContext
import edu.gemini.spModel.rich.shared.immutable._
import edu.gemini.spModel.target.SPTarget
import jsky.app.ot.OT
import jsky.app.ot.ags.BagsManager
import jsky.app.ot.gemini.editor.targetComponent.GuidingFeedback.BagsStatusRow


class GuidingFeedbackEditor extends TelescopePosEditor {
  private val tab: GuidingFeedback.Table = new GuidingFeedback.Table
  def getComponent: Component = tab.peer

  override def edit(ctxOpt: GOption[ObsContext], target: SPTarget, node: ISPNode): Unit = {
    val mt = OT.getMagnitudeTable

    // Construct the rows for the table. Optionally a BAGS row, and then a list of AGS analysis rows.
    val rows = {
      val bagsRow = for {
        n <- Option(node)
        o <- Option(n.getContextObservation)
        s <- BagsManager.instance.bagsStatus(o.getNodeKey)
      } yield BagsStatusRow(s)

      // If the BAGS row is defined, then use it. If not, create the rows corresponding to the analysis.
      bagsRow.fold(
        ctxOpt.asScalaOpt.map(GuidingFeedback.targetAnalysis(_, mt, target)).getOrElse(Nil)
      )(List(_))
    }

    if (rows.isEmpty)
      tab.clear()
    else
      tab.showRows(rows)
  }
}
